sirepo/job_api.py

```diff
@@ -271,8 +271,30 @@ def _run_mode(request_content):
                 request_content.computeJid,
             )
         )
-    return request_content.pkupdate(
-        jobRunMode=j,
-        sbatchCores=m.sbatchCores,
-        sbatchHours=m.sbatchHours,
-    )
+    request_content.jobRunMode = j
+    return _validate_and_add_sbatch_fields(request_content, m)
+
+
+def _validate_and_add_sbatch_fields(request_content, compute_model):
+    def _element_of_list(field, list):
+        def _v(value):
+            assert value in list, \
+                f'value={value} not in enum={list}'
+        _validate_and_add(field, _v)
+
+    def _number(field, *types):
+        def _v(value):
+            assert type(value) in types and value > 0, \
+                f'value={value} expecting a {types} > 0'
+        _validate_and_add(field, _v)
+
+    def _validate_and_add(field, validate):
+        v = getattr(compute_model, field)
+        validate(v)
+        setattr(request_content, field, v)
+
+    _number('sbatchCores', int)
```

Comment:
I think this is too complicated with all these names for one or two lines of code that are self-describing. The error messages are being made generic, when they could be literal and more complete. Here's a rewrite:

```py
m = compute_model
c = request_content
 assert m.sbatchQueue in sirepo.job.NERSC_QUEUES, \
    f'sbatchQueue={m.sbatchQueue} not in NERSC_QUEUES={sirepo.job.NERSC_QUEUES}'
c.sbatchQueue = m.sbatchQueue
for f in 'sbatchCores', 'sbatchHours':
    assert m[f] > 0, f'{f}={m[f]} must be greater than 0'
    c[f] = m[f]
```

The error messages then are self-explanatory. 

sbatchHours can be a float and still tested against 0 in Python. If sbatchHours is not a number, this will fail in a way, but it's an unlikely case, because the GUI is supposed to be sending valid values. We are just protecting against a hacker sending a string where a number should have been sent. It's a balance between producing useful error messages in the important case (number is negative -- which the GUI might protect against).

One Python "bug" is that assertions are not functions. This is annoying. We'd like to able to pass them around, not having to pass them in a wrapped function. In this case, the improved quality of the self-documenting assertions and reduced number of abstractions outweighs the duplication.

At least imiho. :)

Reply:
Much better. Amazing how good I am at creating gobs of code and how good your are at deleting them:)

https://github.com/radiasoft/sirepo/pull/2511/commits/779ddedc7bc17f3c385612be4fee7c6b6a503204